wasm2c.md

```diff
@@ -0,0 +1,332 @@
+      (i32.mul (get_local $x) (call 0 (i32.sub (get_local $x) (i32.const 1))))
+    )
+  )
+)
+```
+
+Save this to `fac.wat`. We can convert this to a `.wasm` file by using the
+`wat2wasm` tool:
+
+```sh
+$ wat2wasm fac.wat -o fac.wasm
+```
+
+We can then convert it to a C source and header by using the `wasm2c` tool:
+
+```sh
+$ wasm2c fac.wasm -o fac.c
+```
+
+## Looking at the generated header, `fac.h`
+
+The generated header file looks something like this:
+
+```c
+#ifndef FAC_H_GENERATED_
+#define FAC_H_GENERATED_
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifndef WASM_RT_INCLUDED_
+#define WASM_RT_INCLUDED_
+
+...
+
+#endif  /* WASM_RT_INCLUDED_ */
+
+extern void WASM_RT_ADD_PREFIX(init)(void);
+
+/* export: 'fac' */
+extern u32 (*WASM_RT_ADD_PREFIX(Z_facZ_ii))(u32);
+#ifdef __cplusplus
+}
+#endif
+
+#endif  /* FAC_H_GENERATED_ */
+```
+
+Let's look at each section. The outer `#ifndef` is standard C boilerplate for a
+header. The `extern "C"` part makes sure to not mangle the symbols if using
+this header in C++.
+
+```c
+#ifndef FAC_H_GENERATED_
+#define FAC_H_GENERATED_
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+...
+
+#ifdef __cplusplus
+}
+#endif
+#endif  /* FAC_H_GENERATED_ */
+```
+
+This `WASM_RT_INCLUDED_` section contains a number of definitions required for
+all WebAssembly modules.
+
+```c
+#ifndef WASM_RT_INCLUDED_
+#define WASM_RT_INCLUDED_
+
+...
+
+#endif  /* WASM_RT_INCLUDED_ */
+```
+
+First we can specify the maximum call depth before trapping. This defaults to
+500:
+
+```c
+#ifndef WASM_RT_MAX_CALL_STACK_DEPTH
+#define WASM_RT_MAX_CALL_STACK_DEPTH 500
+#endif
+```
+
+Next we can specify a module prefix. This is useful if you are using multiple
+modules that may use the same name as an export. Since we only have one module
+here, it's fine to use the default which is an empty prefix:
+
+```c
+#ifndef WASM_RT_MODULE_PREFIX
+#define WASM_RT_MODULE_PREFIX
+#endif
+
+#define WASM_RT_PASTE_(x, y) x ## y
+#define WASM_RT_PASTE(x, y) WASM_RT_PASTE_(x, y)
+#define WASM_RT_ADD_PREFIX(x) WASM_RT_PASTE(WASM_RT_MODULE_PREFIX, x)
+```
+
+Next are some convenient typedefs for integers and floats of fixed sizes:
+
+```c
+typedef uint8_t u8;
+typedef int8_t s8;
+typedef uint16_t u16;
+typedef int16_t s16;
+typedef uint32_t u32;
+typedef int32_t s32;
+typedef uint64_t u64;
+typedef int64_t s64;
+typedef float f32;
+typedef double f64;
+```
+
+Next is the `wasm_rt_trap_t` enum, which is used to give the reason a trap
+occurred.
+
+```c
+typedef enum {
+  WASM_RT_TRAP_NONE,
+  WASM_RT_TRAP_OOB,
+  WASM_RT_TRAP_INT_OVERFLOW,
+  WASM_RT_TRAP_DIV_BY_ZERO,
+  WASM_RT_TRAP_INVALID_CONVERSION,
+  WASM_RT_TRAP_UNREACHABLE,
+  WASM_RT_TRAP_CALL_INDIRECT,
+  WASM_RT_TRAP_EXHAUSTION,
+} wasm_rt_trap_t;
+```
+
+Next is the `wasm_rt_type_t` enum, which is used for specifying function
+signatures. The four WebAssembly value types are included:
+
+```c
+typedef enum {
+  WASM_RT_I32,
+  WASM_RT_I64,
+  WASM_RT_F32,
+  WASM_RT_F64,
+} wasm_rt_type_t;
```

Comment:
Not obvious why you'd want these at runtime (and quite a bit of the other data in this section as well). I'd imagine you'd simply generate a cast at the call site affert you `assert` the two function indices are the the same.

Reply:
You only need them if you're linking multiple modules. Since two modules can both define the same function signature with different indexes, the embedder needs to determine whether they're the same.

Perhaps there should be an option to say: I only need one module, don't give me this other stuff.